### ISSUE TITLE
Inconsistent polygon area computation

### ISSUE THREAD

**Issue by mcara** [spacetelescope/sphere#73]
I would have expected the **sign** of the polygon area to depend on the orientation of the polygon. However, it seems that this is not the case with `SphericalPolygon` in the `sphere` package: when polygon's orientation is reversed, its area (as computed by `SphericalPolygon.area()`) is `expected_area`+`4*pi`, where `expected_area` is negative of the "normally" oriented polygon.

Here is an example:

```python
>>> from sphere.polygon import SphericalPolygon
>>> import numpy as np
>>> ra = [-158.60875065205286, -158.60842667643962, -158.60835235580953, 
... -158.60867652448502, -158.60875065205286]
>>> dec = [-43.003818432626232, -43.003746119755469, -43.00393669833641, 
... -43.004008935939986, -43.003818432626232]
>>> s1 = SphericalPolygon.from_radec(ra, dec)
>>> ra.reverse()
>>> dec.reverse()
>>> s2 = SphericalPolygon.from_radec(ra, dec)
>>> s1.area()
1.4950707338812208e-11
>>> s2.area()
12.566370614344223
>>> s2.area() - 4.0 * np.pi
-1.4949819160392508e-11
# this is what I was expecting s2.area() to return
```

Also, please notice that the relative difference between the area computed with one orientation and the area that I was expecting for the other orientation is about 0.006%. This seems a little too high considering that the vertices of the polygons are identical and only their ordering is different.

**Comment by pllim** [comment]
Should area even have signs? Negative area does not make sense to me.

**Comment by mcara** [comment]
Last sentence on this page http://mathworld.wolfram.com/PolygonArea.html explains the usual convention. For my particular problem I need to take `|Area|` but I can imagine cases when sign can be used to infer polygon's orientation.

**Comment by bernie-simon** [comment]
I think you're both right. Normally you expect an area to be positive. But there are times when having a signed area is useful. Probably I will add a new method, signed_area() which will do the work and area() will take its abolute value. If you have a better name for signed_area(), please let me know.

**Comment by mcara** [comment]
I think we should use a signed area convention for 'area' of a polygon. Users can call 'abs' themselves or you could have 'absolute_area' for those unable to do so.

The fundamental issue is that the area of the reversed polygon is wrong!

**Comment by bernie-simon** [comment]
The problem with changing area() to return a signed value is that it's an interface change and if it is made, code that uses it will break and need to be modified. This is not the way to get people to use this package. Since it is not an error to always return a positive area and that is what the majority of users expect, I prefer to  leave area() alone and add a new method signed_area().

**Comment by mcara** [comment]
@bernie-simon wrote: _"The problem with changing area() to return a signed value is that it's an interface change and if it is made, code that uses it will break and need to be modified."_

I disagree with the above statement: Right now no code that uses `area()` on clockwise oriented polygons should work. For example, instead of returning `-1.494982e-11`, current API returns `12.566`. So, if what you say is true, then we should continue returning the incorrect `4*pi-true_area` just because if you will start returning the true value (even the positive `+1.494982e-11`) any existing code will crash.

Again, returning positive area would be counter to commonly used convention. Unfortunately I cannot check how `SciPy` works (my version is older) but if yoy have scipy v16 (or v16.1), try to compute the area of a convex hull for a CCW and CW oriented polygons: http://scipy.github.io/devdocs/generated/scipy.spatial.ConvexHull.html

Even more important is that `Shapely` and may other GIS applications return signed area.

However, if a signed area is really objectionable to you, consider having a function such as "orientation()" or "is_ccw()" instead of creating `signed_area()` besides (absolute) `area()`.

**Comment by mcara** [comment]
CC @stsci-hack [due to this bug tweakreg for HST and JWST will not function correctly as polygons areas vary *widely*]

**Comment by bernie-simon** [comment]
Trying to figure out the best way to solve this problem. Would it cause you or other people a problem if I check the area of the polygon when creating it and if it is between two and four pi ("negative") and the inside point is inside the polygon, reversing the order of the points so that the area is between zero and two pi?

**Comment by mcara** [comment]
I do not think this is a sound approach for the following reasons:

1. The polygon object **should not** change vertex ordering. This is plain wrong.
2. In addition, as I have mentioned above, having signed area is a plus.
3. If I understand correctly, you want to patch a buggy code by computing area as
```
new_area = old_area if old_area < pi else 2*pi - old_area
```
This looks like a working patch as long as a polygon cannot have true area > pi. However, this patch seems to be avoiding a real problem with the code.

**Comment by bernie-simon** [comment]
The area computation is now signed. Polygon points are not re-ordered. The computation of area is done by Girard's theorem and I have examined it to make sure that it is correct, so I have no doubts the correct formula is being used. The only remaining problem is that I have not handled the case where the inside point is "outside" the polygon. (In the larger portion of the sphere.) In that case, the area really should be between 2 and 4 pi.

**Comment by mcara** [comment]
How can an inside point become an "outside"? Could this be because of the way shpere creates "unions"? In past it used to have a single polygon that was made up from multiple polygons that were connected through "cuts". Later @mdboom started using lists of single polygons to represent unions.

One issue could be that when computing the "inner" point of two adjacent rectangles (like two chips in an ACS image) but that have a small gap between them, this inner points (if it is computed as the average of vertex coordinates) fall in the gap between the polygons making up an union and thus it becomes an "outer" point???

Also another suggestion (that may or may not be related to the current issue) is that when an union is created, these individual polygons that make up an union be oriented in the same way as the first polygon. Otherwise it is easy to get the area wrong if some polygons contribute with positive areas while other polygons contribute with negative areas to the total area of a union polygon. Alternatively, instead of re-orienting polygons, area of the union polygon could be computed as:
```
total=sign(area[0]) * (abs(area[0]) + abs(area[1]) + ...)
```
I am not sure this is a problem in the current code - just a thought to check.

**Comment by mcara** [comment]
> _"The only remaining problem is that I have not handled the case where the inside point is "outside" the polygon. (In the larger portion of the sphere.) In that case, the area really should be between 2 and 4 pi."_

If that is the case, then why did you close this issue??? This was the main point of the original issue! The signed area problem was rised by @pllim for whom negative areas make no sense.

**Comment by bernie-simon** [comment]
An inside point can lie on the "outside" because the user chooses the inside point when creating a polygon. Multiple disjoint polygons are made of single polygons, each which has an inside point.There seems to be no perfect solution to computing the area of a spherical polygon, but returning negative values seems preferable to returning values between 2 and 4 pi, which I think most users find unintuitive. Users like @plim who always want a positive area can take the absolute value. It is a bit harder to test for and correct the area when it falls between 2 and 4 pi.

The issue was closed because a fix has been committed along the lines explained here.

**Comment by mcara** [comment]
It seems that this was a known issue: https://github.com/spacetelescope/sphere/issues/44#issuecomment-55038001

It looks to me that something is fundamentally screwed up with the sphere's logic if one needs to re-orient polygons...

**Comment by bernie-simon** [comment]
Girard's theorem returns a value between 2 and 4 pi if you traverse the polygon in a counter-clockwise order. It's as simple as that.

**Comment by mcara** [comment]
I apologize that things are not as simple as they are for you.

Are you implying that true area of a polygon cannot be between 2pi and 4pi? That is imagine a very large polygon covering almost the entire sphere except for a small gap at the "pole". I would expect the area of such a polygon be > 2pi. For specificity, let's assume that the true area of large polygon is 3pi. Are you going to return -pi now instead of 3pi?

**Comment by bernie-simon** [comment]
That is the case where the inside point is on the "outside," which I do not handle now, as I have said. Will try to find the time to handle it later, but for now there are more pressing issues.

**Comment by mcara** [comment]
@bernie-simon Was this issue fixed? If yes, could you link this issue with a specific pull request? [Also, in the future it would be helpful to link issues with pull requests that fixes those issues]

**Comment by mcara** [comment]
My tests indicate that this issue has not been solved and that the area of a small polygon is still about 4*pi:

```puthon
>>> from spherical_geometry.polygon import SphericalPolygon as SP
>>> from astropy import wcs
>>> import numpy as np
>>> fwi = wcs.WCS()
>>> fwi.wcs.pc = [[-1.666667E-05, 0], [0, 1.666667E-05]]
>>> fwi.wcs.cdelt = [1, 1]
>>> fwi.wcs.crpix = [4100.5, -8157.5]
>>> fwi.wcs.crval = [150.1163213, 2.200973097]
>>> fwi.wcs.ctype = ['RA---TAN', 'DEC--TAN']
>>> fwi.wcs.radesys = 'ICRS'
>>> fwi.wcs.set()
>>> pts1 = [[0, 0], [0, 100], [100, 100], [100, 0], [0, 0]]
>>> rd1 = fwi.wcs_pix2world(pts1, 1)
>>> p1 = SP.from_radec(rd1[:, 0], rd1[:, 1])
>>> p1.area()
12.566370613513023

**Comment by bernie-simon** [comment]
The fix was never pulled into the spacetelescope repository. I intend to do that as soon as the tests run and I will tag a new version, 1.0.7.

**Comment by mcara** [comment]
> I intend to do that as soon as the tests run and I will tag a new version, 1.0.7.
That's fine. However, please do not close unsolved issues.

**Comment by bernie-simon** [comment]
The change has been merged in pull request #77. I will be tagging the code as 1.0.7 in a few minutes. Please let me know if you continue to have problems. I did not encounter any when I ran your test case.

**Comment by bernie-simon** [comment]
The fix has been tagged as 1.0.7 and I have verified that the fix is on this repository.
